src/RovesODrive.cpp

```diff
@@ -1,425 +1,486 @@
+void writeODriveConfig(HardwareSerial* mySerial, bool write_read, char* id, char* param, uint8_t axis)
 {
-	int value = 0;
-	int i = 0;
-	while(input[i] != '\0')
-	{
-		value *= 10;
-		value += input[i] - '0';
-		i++;
-	}
-	return(value);
+	char output[MAX_STRING_CHARS];
+
+	sprintf(output, "%s ", (write_read == WRITE)? "w":"r");
+
+	sprintf(output, "%s%s%d.", output, "axis", axis);
+
+	sprintf(output, "%s%s %s\n", output, id, param);
+	mySerial->write(output);
+	return;
+
 }
 
-float charToFloat(char input[])
+void writeODriveConfig(HardwareSerial* mySerial, bool write_read, char* id, float& param, uint8_t axis)
 {
-	float value = 0;
-	int i = 0;
-	while(input[i] != '.')
-	{
-		value *= 10;
-		value += input[i] - '0';
-		i++;
-	}
-	int j = i++;
-	while(input[i] != '\0')
+	char output[MAX_STRING_CHARS];
+
+	char data[20];
+
+	floatToChar(data, param, 4);
+
+	sprintf(output, "%s ", (write_read == WRITE)? "w":"r");
+
+	sprintf(output, "%s%s%d.", output, "axis", axis);
+
+	sprintf(output, "%s%s %s\n", output, id, data);
+	mySerial->write(output);
+	return;
+
+}
+
+
+void writeODriveCommand(HardwareSerial* mySerial, char* id, char* param1, char* param2, char* param3, uint8_t axis) //TODO: add two additional parameters for commands
+{
+	char output[MAX_STRING_CHARS];
+
+	sprintf(output, "%s", id);
+
+	sprintf(output, "%s %d", output, axis);
+
+	sprintf(output, "%s %s", output, param1);
+
+	sprintf(output, "%s %s", output, param2);
+
+	sprintf(output, "%s %s \n", output, param3);
+
+	mySerial->write(output);
+}
+
+void RovesODrive::begin(HardwareSerial* mySerial)
+{
+	//init serial and motors
+	m_serial = mySerial;
+	left.motor_number = 0;
+	left.m_serial = mySerial;
+	right.motor_number = 1;
+	right.m_serial = mySerial;
+
+	//startup serial with proper baud rate
+	m_serial->begin(115200);
+	delay(10);
+	m_serial->write("\n");
+}
+
+String RovesODriveMotor::getSerial()
+{
+	String str = "";
+    static const unsigned long timeout = 1000;
+    unsigned long timeout_start = millis();
+    while( true ) 
 	{
-		value *= 10;
-		value += input[i] - '0';
-		i++;
-	}
-	return(value/(10*(i-j)));
+        while (!m_serial->available()) {
+            if (millis() - timeout_start >= timeout) {
+                return str;
+            }
+        }
+        char c = m_serial->read();
+        if (c == '\n')
+            break;
+        str += c;
+    }
+    return str;
+}
+
+void RovesODriveMotor::writeState(Axis_State state) 
+{
+	char data[10];
+	intToChar(data, state);
+	writeODriveConfig(m_serial, WRITE, WRITE_CURRENT_STATE, data, motor_number);
 }
 
-bool charToBool(char input[])
+Axis_State RovesODriveMotor::readState()
 {
-	return(1);
+	String state = ""; 
+	writeODriveConfig(m_serial, READ, READ_CURRENT_STATE, "", motor_number);
+	state = getSerial();
+	return (Axis_State)state.toInt();
 }
 
-void intToChar(char* output, int value)
+void RovesODriveMotor::writeControlMode(Control_Mode mode)
 {
-	sprintf(output, "%d", value);
+	char data[2];
```

Comment:
Go through all implicit data type -> character array conversions and make sure they are all the good sizes.